ui/src/components/credentials/CredentialSubjectForm.tsx

```diff
@@ -0,0 +1,109 @@
+import { Card, Space, Typography } from "antd";
+import { Fragment, useRef } from "react";
+
+import { AttributeBreadcrumb } from "src/components/credentials/AttributeBreadcrumb";
+import { Boolean } from "src/components/credentials/attributes/Boolean";
+import { Number } from "src/components/credentials/attributes/Number";
+import { String } from "src/components/credentials/attributes/String";
+import { Attribute, ObjectAttribute } from "src/domain";
+
+function AnyAttribute({
+  attribute,
+  parents,
+}: {
+  attribute: Attribute;
+  parents: ObjectAttribute[];
+}): JSX.Element {
+  switch (attribute.type) {
+    case "boolean": {
+      return <Boolean attribute={attribute} parents={parents} />;
+    }
+    case "number":
+    case "integer": {
+      return <Number attribute={attribute} parents={parents} />;
+    }
+    case "string": {
+      return <String attribute={attribute} parents={parents} />;
+    }
+    case "null": {
+      return (
+        // ToDo: PID-587
+        <Typography.Text>Null attributes are not yet supported ({attribute.name})</Typography.Text>
+      );
+    }
+    case "multi": {
+      return (
+        // ToDo: PID-543
+        <Typography.Text>Multi attributes are not yet supported ({attribute.name})</Typography.Text>
+      );
+    }
+    case "array": {
+      return (
+        // ToDo: PID-589
+        <Typography.Text>Array attributes are not yet supported ({attribute.name})</Typography.Text>
+      );
+    }
+    case "object": {
+      return (
+        <CredentialSubjectForm
+          attributes={attribute.schema.properties || []}
+          parents={[...parents, attribute]}
+          space={false}
+        />
+      );
+    }
+  }
+}
+
+export function CredentialSubjectForm({
+  attributes,
+  parents = [],
+  space = true,
+}: {
+  attributes: Attribute[];
+  parents?: ObjectAttribute[];
+  space?: boolean;
+}) {
+  const shouldShowBreadcrumb = useRef<boolean>(true);
+  const form = [...attributes]
+    .sort((a, b) => (a.type !== "object" && b.type !== "object" ? 0 : a.type === "object" ? 1 : -1))
+    .map((attribute: Attribute) => {
+      const showBreadcrumb =
+        attribute.type !== "object" && parents.length > 1 && shouldShowBreadcrumb.current;
```

Comment:
Did you test the [Family tree](https://schemas.serto.id/schema/family-tree) schema?

Reply:
My question is related to a less kludgy (more data-driven) way than abusing refs. Doing the following renders it identically unless I'm missing something;

![Screenshot from 2023-04-18 12-20-18](https://user-images.githubusercontent.com/2691584/232748278-6fb81343-374f-47b4-bf06-c2a9807176eb.png)

Reply:
My initial attempt was to use the condition `index === 0` and it was not working but I don't recall the reason. It was surely with a previous implementation because now it works as you point out.

Love removing this ref! Done!